13.08.2023

jakogri opened the pull request:
Задачи:
1. Добавить предупреждение перед созданием компании.
2. Сделать новую компанию невидимой до решения администрации.
3. Послать емайл о создании компании на адрес администратора. 
 Решения:
1. Ввёл новое модальное окно в модуль CreateCampaqign. Оно первым  возникает при создании компании.В файлы lang ввёл новые ключи с полями  из предупреждения.
2. При создании компании ключ active по умолчанию имеет значение false.
3. Изменил модули server.js и serverLib.js. В первый ввёл оператор post  с адресом '/api/email/sendemail'. Во второй - функцию handleSendEmail,  которая и посылает емайл. 
4. В базу данных добавил коллекцию email_config, в документах которой и  должны содержаться данные для отсылки сообщений

jakogri committed:
13.08.2023
Задачи:
1. Добавить предупреждение перед созданием компании.
2. Сделать новую компанию невидимой до решения администрации.
3. Послать емайл о создании компании на адрес администратора.
Решения:
1. Ввеёл новое модальное окно в модуль CreateCampaqign. Оно первым 
возникает при создании компании.В файлы lang ввёл новые ключи с полями 
из предупреждения.
2. При создании компании ключ active по умолчанию имеет значение false.
3. Изменил модули server.js и serverLib.js. В первый ввёл оператор post 
с адресом '/api/email/sendemail'. Во второй - функцию handleSendEmail, 
которая и посылает емайл. 
4. В базу данных добавил коллекцию email_config, в документах которой и 
должны содержаться данные для отсылки сообщений

grishick commented on src/lang/translationEN.json:
```diff
@@ -17,6 +17,7 @@
     "raised" : " raised of ",
     "goal" : "goal",
     "campaignCreateSuccess": "Congratulations! Your fundraiser is live on the blockchain",
+    "campaignwWillBePublished": "Thank you for your application. Your campaign will be published at our platform after verification process, which takes 1-3 working days",
```
поменяй пожалуйста на "Thank you for your application. Your campaign is being reviewed and will be published within 3 business days"

grishick commented on server/serverLib.js:
```diff
@@ -43,6 +45,32 @@ class ServerLib {
+    async handleSendEmail(req, res, Sentry, DB){
+       try{
+        const emailCollection = await DB.collection('email_config');
+        let result = await emailCollection.findOne({"_id" : req.body.key});  
```
вот тут я не понимаю. Почему мы ищем в emailCollection по _id пришедшему в запросе? Какая связь между emailCollection и req.body.key?

grishick replied:
давай лучше положим все email настройки в уже существующую коллекцию `global_configs` и _id не нужно передавать из UI, а лучше задать в server.js

grishick commented on server/server.js:
```diff
@@ -176,6 +176,11 @@ APP.post('/api/donate/adddanate', async (req, res) => {
 });
 
+APP.post('/api/email/sendemail', (req, res) => {
+        const DB = CLIENT.db(DBNAME);
```
поскольку этот URL не требует логина, то злоумышленник может написать скрипт, который будет делать бесконечные запросы на этот адрес и таким образом зафлудит нашу почту. Нужно, чтобы этот URL был доступен только авторизованному пользователю

grishick commented on src/components/CreateCampaign.js:
```diff
@@ -222,19 +223,25 @@ class CreateCampaign extends React.Component {
+             ". Сampaign name " + campaignData.org["default"] + ". Сampaign ID - " +  campaignData.address + ". Сampaign beneficiary - " + 
+             campaignData.beneficiaryId + ".";
+            let data = {key : "New Campaign Alert",text: text};
+            axios.post('/api/email/sendemail', data, {headers: {"Content-Type": "application/json"}});
```
вместо того, чтобы вызывать посылку email отдельным POST запросом из веб страницы, мне кажется, лучше вызвать функцию посылки email прямо в `server.js`  или `serverLib::handleAddCampaign`

grishick reviewed (changes requested):
Отличное описание PR! Я оставил несколько комментариев в файлах

jakogri committed:
17.09.2023
Задачи:
1. Уйти от отдельного запроса на сервер с целью вызова функции посылки 
емайл сообщения.
2. Избежать возможности отсылки емайл сообщения неавторизированным 
пользователем.
3. Убрать из параметров запроса на сервер значение ключа для вызова 
записи с параметрами для посылки емайл.
Решения:
1.Изменил процедуру запроса Post на сервере. Сперва из swerverlib вызваю 
функцию записи новой компании в базу данных. По её удачному окончанию 
вызываю из serverlib функцию отсылки емайл.Из неё же возвращаю на 
клиента сообщение об удачном окончании, либо об ошибке. В случае же если 
запись в базу данных заканчивается ошибкой, то сообщение оней возвращаю 
на клиента из запроса Post, не вызывая функции отсылки емайл.

grishick reviewed (changes requested):
Всё отлично. Есть только две маленькие просьбы: 
1. пусть _id (адрес) кампании в email будет весь в нижнем регистре, чтобы можно было скопировать из email прямо в поисковый запрос в mongoDB {_id:"0x46f65704bf702dc18ee3f7d4e53b8a99eafc55b8"}
2. также, добавь в текст email поле key

jakogri committed:
Задача:
1) Для текста в емайл перевести адресса в нижний регистр и добавить 
значение ключа компании.
Решение:
1) Внёс соответствующие изменения в формировании константы text в 
операцции popst на сервере, вызывающей функцию отсылки емайл.

grishick commented on server/server.js:
```diff
@@ -176,6 +182,11 @@ APP.post('/api/donate/adddanate', async (req, res) => {
     }
 });
 
+APP.post('/api/email/sendemail', (req, res) => {
```
этот handler кажется больше не используется. Можешь убрать?

grishick reviewed (changes requested):
убери пожалуйста не используемый handler /api/email/sendemail

jakogri committed:
19.09.2023 12:54
Убрал из server.js неиспользуемый handler /api/email/sendemail
